<h2>Objectives:</h2>
<li>Do "<b>Ask HN</b>" or "<b>Show HN</b>" receive more comments on average?</li>
<li>Do posts created at a certain time receive more comments on average?</li>
<p>Y Hacker News dataset on <b>Kaggle</b> can be found <a href="https://www.kaggle.com/hacker-news/hacker-news-posts">here</a>.

In [1]:
import csv
open_file = open("hacker_news.csv")
hn = list(csv.reader(open_file))
print(hn[:5])    # prints out first 5 rows in a messy list format below
headers = hn[0]  # assign header/column names to 'headers' variable
hn = hn[1:]      # removes header so we don't have to continuously slice out header

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [2]:
print(headers)
# column names description below

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


<h3>Column Names:</h3>
<ul><b>id</b> - The unique identifier from Hacker News for the post</ul>
<ul><b>title</b> - The title of the post</ul>
<ul><b>url</b> - The URL that the posts links to, if it the post has a URL</ul>
<ul><b>num_points</b> - The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes</ul>
<ul><b>num_comments</b> - The number of comments that were made on the post</ul>
<ul><b>author</b> - The username of the person who submitted the post</ul>
<ul><b>created_at</b> - The date and time at which the post was submitted</ul></p>
<br>
<h3>Users submit "Ask HN" posts to ask the Hacker News community a specific question.</h3><br>
<p><b>Examples:</p>
<li>Ask HN: How to improve my personal website?</li>
<li>Ask HN: Am I the only one outraged by Twitter shutting down share counts?</li>
<li>Ask HN: Aby recent changes to CSS that broke mobile?</li>
<h3>Users submit "Show HN" posts to show the Hacker News community a project, product, or just generally something interesting.</h3><br>
<p><b>Examples:</p>
<li>Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform'</li>
<li>Show HN: Something pointless I made</li>
<li>Show HN: Shanhu.io, a programming playground powered by e8vm</li>

In [3]:
ask_posts = []    # list to store 'ask hn' posts
show_posts = []   # list to store 'show hn' posts
other_posts = []  # other irrevelant posts

# loop through each row in dataset
for row in hn:
    # store second column's value into title
    title = row[1]
    
    if title.lower().startswith("ask hn"):
        # if value starts with 'ask hn', append to ask_posts
        ask_posts.append(row)
    elif title.lower().startswith("show hn"):
        # if value starts with 'show hn', append to show_posts
        show_posts.append(row)
    else:
        # if it is neither, append to other_posts
        other_posts.append(row)

print(len(ask_posts))    # prints number of ask hn posts
print(len(show_posts))   # prints number of show hn posts
print(len(other_posts))  # prints number of other posts

1744
1162
17194


In [14]:
print(ask_posts[0])    # printing to see what each row consists of, so far contains all strings

['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55']


In [16]:
print(headers)    # for referencing what each column means

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


<li>'id' : '12296411'</li>
<li>'title' : 'Ask HN: How to improve my personal website?'</li>
<li>'url' : ''</li>
<li>'num_points' : '2'</li>
<li>'num_comments' : '6'</li>
<li>'author' : 'ahmedbaracat'</li>
<li>'created_at' : '8/16/2016 9:55'</li>

In [26]:
total_ask_comments = 0    # looking to see how many num_comments there are, counter begins at 0

# loop through each post in new ask_posts list consisting of 'ask hn' comments
for post in ask_posts:
    total_ask_comments += int(post[4])    # convert num_comments to int and adding to total
    
avg_ask_comments = total_ask_comments/len(ask_posts)    # finding average of num_comments
print("'Ask HN' num_comments average: " + str(avg_ask_comments))

'Ask HN' num_comments average: 14.038417431192661


In [27]:
total_show_comments = 0

for post in show_posts:
    total_show_comments += int(post[4])
    
avg_show_comments = total_show_comments/len(show_posts)
print("'Show HN' num_comments average: " + str(avg_show_comments))

'Show HN' num_comments average: 10.31669535283993


By storing both ask comments and show comments into two different lists then comparing them by average count:<br>
<b>'Ask HN' posts receive more comments on average although there are much more other posts.</b><br><br>
That answers our first objective.
Now we move on to answer our second objective, which is:<br>
<b>Do posts created at a certain time receive more comments on average?</b>

In [18]:
import datetime as dt    # import the datetime module

result_list = []         # create an empty list to store time and number of comments

for post in ask_posts:
    result_list.append([post[6], int(post[4])])
    # we append the time followed by number of comments to the result list

comments_by_hour = {}    # dictionary to store comments by the hour
counts_by_hour = {}      # dictionary to store counts by the hour

# predetermining format for date/time
date_format = "%m/%d/%Y %H:%M"

# now looping through each row in the result list
for each_row in result_list:
    date = each_row[0]    # assigning the first element in each row as the date
    comment = each_row[1] # assigning the second element as the comment
    
    # using datetime module, convert the date with format from above into 24-hrs
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    
    # if the hour is in the counts, assign the comment value to the time key of comments by the hour
    # then increment the time key in counts by the hour by 1
    # otherwise assign the comment to time in comments by the hour
    # then assign the time key in counts by the hour to 1 since the key doesn't exist in dictionary
    if time in counts_by_hour:
        comments_by_hour[time] += comment
        counts_by_hour[time] += 1
    else:
        comments_by_hour[time] = comment
        counts_by_hour[time] = 1

comments_by_hour

{'00': 447,
 '01': 683,
 '02': 1381,
 '03': 421,
 '04': 337,
 '05': 464,
 '06': 397,
 '07': 267,
 '08': 492,
 '09': 251,
 '10': 793,
 '11': 641,
 '12': 687,
 '13': 1253,
 '14': 1416,
 '15': 4477,
 '16': 1814,
 '17': 1146,
 '18': 1439,
 '19': 1188,
 '20': 1722,
 '21': 1745,
 '22': 479,
 '23': 543}

In [32]:
counts_by_hour

{'00': 55,
 '01': 60,
 '02': 58,
 '03': 54,
 '04': 47,
 '05': 46,
 '06': 44,
 '07': 34,
 '08': 48,
 '09': 45,
 '10': 59,
 '11': 58,
 '12': 73,
 '13': 85,
 '14': 107,
 '15': 116,
 '16': 108,
 '17': 100,
 '18': 109,
 '19': 110,
 '20': 80,
 '21': 109,
 '22': 71,
 '23': 68}

In [7]:
avg_by_hour = []    # empty list for holding average amount of comments per hour from 'Ask HN'

# loop through comments by the hour
for avg in comments_by_hour:
    # append the key/time followed by the average of comments by the hour as a list of lists
    # to do so, divide hourly comments by hourly counts
    avg_by_hour.append([avg, comments_by_hour[avg]/counts_by_hour[avg]])
    
avg_by_hour    # prints out an unsorted list of lists in the format hour, average comments

[['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['14', 13.233644859813085],
 ['19', 10.8],
 ['00', 8.127272727272727],
 ['22', 6.746478873239437],
 ['17', 11.46],
 ['07', 7.852941176470588],
 ['01', 11.383333333333333],
 ['16', 16.796296296296298],
 ['09', 5.5777777777777775],
 ['08', 10.25],
 ['02', 23.810344827586206],
 ['20', 21.525],
 ['11', 11.051724137931034],
 ['23', 7.985294117647059],
 ['13', 14.741176470588234],
 ['12', 9.41095890410959],
 ['21', 16.009174311926607],
 ['04', 7.170212765957447],
 ['15', 38.5948275862069],
 ['10', 13.440677966101696],
 ['05', 10.08695652173913],
 ['06', 9.022727272727273]]

In [29]:
# swapping the hour and average number of comments
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
    
swap_avg_by_hour

[[13.20183486238532, '18'],
 [7.796296296296297, '03'],
 [13.233644859813085, '14'],
 [10.8, '19'],
 [8.127272727272727, '00'],
 [6.746478873239437, '22'],
 [11.46, '17'],
 [7.852941176470588, '07'],
 [11.383333333333333, '01'],
 [16.796296296296298, '16'],
 [5.5777777777777775, '09'],
 [10.25, '08'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [11.051724137931034, '11'],
 [7.985294117647059, '23'],
 [14.741176470588234, '13'],
 [9.41095890410959, '12'],
 [16.009174311926607, '21'],
 [7.170212765957447, '04'],
 [38.5948275862069, '15'],
 [13.440677966101696, '10'],
 [10.08695652173913, '05'],
 [9.022727272727273, '06']]

In [30]:
# reverse sort to show the descending average of comments by the hour
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
sorted_swap

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [31]:
print("Top 5 Hours for Ask Posts Comments")

for avg, hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    )

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


<h3>Do posts created at a certain time receive more comments on average?</h3>
<p>
Above shows that 3PM averages the most comments per post with an average of 38.59.<br>
Followed by 2AM, 8PM, 4PM and 9PM.
</p>